Build Fails

## Description
Build fails with 

```
➜  terraform-provider-ionoscloud git:(master) make build
==> Checking that code complies with gofmt requirements...
gofmt needs running on the following files:
./ionoscloud/data_source_k8s_cluster.go
You can use the command: `make fmt` to reformat code.
make: *** [GNUmakefile:32: fmtcheck] Error 1
```

After fixing formatting there are multiple errors:
```
➜  terraform-provider-ionoscloud git:(master) ✗ make build
==> Checking that code complies with gofmt requirements...
go install
ionoscloud/provider.go:6:2: cannot find package "." in:
        /home/mneverov/go/src/github.com/ionos-cloud/terraform-provider-ionoscloud/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/diag
ionoscloud/provider.go:14:2: cannot find package "." in:
        /home/mneverov/go/src/github.com/ionos-cloud/terraform-provider-ionoscloud/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource
ionoscloud/data_source_datacenter.go:9:2: cannot find package "." in:
        /home/mneverov/go/src/github.com/ionos-cloud/terraform-provider-ionoscloud/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema
ionoscloud/data_source_k8s_node_pool.go:8:2: cannot find package "." in:
        /home/mneverov/go/src/github.com/ionos-cloud/terraform-provider-ionoscloud/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation
ionoscloud/config.go:5:2: cannot find package "." in:
        /home/mneverov/go/src/github.com/ionos-cloud/terraform-provider-ionoscloud/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/meta
main.go:6:2: cannot find package "." in:
        /home/mneverov/go/src/github.com/ionos-cloud/terraform-provider-ionoscloud/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/plugin
ionoscloud/config.go:6:2: cannot find package "." in:
        /home/mneverov/go/src/github.com/ionos-cloud/terraform-provider-ionoscloud/vendor/github.com/ionos-cloud/sdk-go/v5
ionoscloud/data_source_k8s_cluster.go:10:2: cannot find package "." in:
        /home/mneverov/go/src/github.com/ionos-cloud/terraform-provider-ionoscloud/vendor/gopkg.in/yaml.v3
make: *** [GNUmakefile:9: build] Error 1
```

Also, the `Developing the Provider` section mentions `make bin`, which should probably be replaced with `make build`.

## Expected behavior
Build is successful.

Provider version: 
Latest master.

OS:
Ubuntu 20.04

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+## 5.2.9
+
+- issue #72 - fixed find volume image by name
+
 ## 5.2.8
 
 - rollback to the node pool behaviour before the fix of issue #71

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,7 @@
 ## 5.2.9
 
 - issue #72 - fixed find volume image by name
+- error messages for immutable node pool attributes
 
 ## 5.2.8
 

--- a/README.md
+++ b/README.md
@@ -69,7 +69,7 @@ If you wish to work on the provider, you'll first need [Go](http://www.golang.or
 To compile the provider, run `make build`. This will build the provider and put the provider binary in the `$GOPATH/bin` directory.
 
 ```sh
-$ make bin
+$ make build
 ...
 $ $GOPATH/bin/terraform-provider-ionoscloud
 ...

--- a/docs/resources/k8s_node_pool.md
+++ b/docs/resources/k8s_node_pool.md
@@ -43,7 +43,7 @@ resource "ionoscloud_k8s_node_pool" "demo" {
 
 The following arguments are supported:
 
-- `name` - (Required)[string] The name of the Kubernetes Cluster.
+- `name` - (Required)[string] The name of the Kubernetes Cluster. *This attribute is immutable*.
 - `k8s_version` - (Optional)[string] The desired Kubernetes Version. for supported values, please check the API documentation. The provider will ignore changes of patch level.
 - `auto_scaling` - (Optional)[string] Wether the Node Pool should autoscale. For more details, please check the API documentation
   - `min_node_count` - (Required)[int] The minimum number of worker nodes the node pool can scale down to. Should be less than max_node_count
@@ -54,13 +54,13 @@ The following arguments are supported:
   - `day_of_the_week` - (Required)[string] Day of the week when maintenance is allowed
 - `datacenter_id` - (Required)[string] A Datacenter's UUID
 - `k8s_cluster_id`- (Required)[string] A k8s cluster's UUID
-- `cpu_family` - (Required)[string] The desired CPU Family - See the API documentation for more information
-- `availability_zone` - (Required)[string] - The desired Compute availability zone - See the API documentation for more information
-- `storage_type` -(Required)[string] - The desired storage type - SSD/HDD
+- `cpu_family` - (Required)[string] The desired CPU Family - See the API documentation for more information. *This attribute is immutable.*
+- `availability_zone` - (Required)[string] - The desired Compute availability zone - See the API documentation for more information. *This attribute is immutable.*
+- `storage_type` -(Required)[string] - The desired storage type - SSD/HDD. *This attribute is immutable.*
 - `node_count` -(Required)[int] - The desired number of nodes in the node pool
-- `cores_count` -(Required)[int] - The CPU cores count for each node of the node pool
-- `ram_size` -(Required)[int] - The desired amount of RAM, in MB
-- `storage_size` -(Required)[int] - The desired amount of storage for each node, in GB
+- `cores_count` -(Required)[int] - The CPU cores count for each node of the node pool. *This attribute is immutable.*
+- `ram_size` -(Required)[int] - The desired amount of RAM, in MB. *This attribute is immutable.*
+- `storage_size` -(Required)[int] - The desired amount of storage for each node, in GB. *This attribute is immutable.*
 - `public_ips` - (Optional)[list] A list of public IPs associated with the node pool; must have at least `node_count + 1` elements;  
 
 ## Import

--- a/ionoscloud/data_source_k8s_cluster.go
+++ b/ionoscloud/data_source_k8s_cluster.go
@@ -10,27 +10,27 @@ import (
 	"gopkg.in/yaml.v3"
 )
 
-type KubeConfig struct{
-	ApiVersion string			`yaml:"apiVersion"`
-	Clusters []struct{
-		Name string
-		Cluster struct{
-			CaData string		`yaml:"certificate-authority-data"`
+type KubeConfig struct {
+	ApiVersion string `yaml:"apiVersion"`
+	Clusters   []struct {
+		Name    string
+		Cluster struct {
+			CaData string `yaml:"certificate-authority-data"`
 			Server string
 		}
 	}
-	Contexts []struct{
-		Name string
-		Context struct{
+	Contexts []struct {
+		Name    string
+		Context struct {
 			Cluster string
-			User string
+			User    string
 		}
 	}
-	CurrentContext string		        `yaml:"current-context"`
-	Kind string
-	Users []struct{
+	CurrentContext string `yaml:"current-context"`
+	Kind           string
+	Users          []struct {
 		Name string
-		User struct{
+		User struct {
 			Token string
 		}
 	}
@@ -104,40 +104,38 @@ func dataSourceK8sCluster() *schema.Resource {
 				Computed: true,
 			},
 			"config": {
-				Type: schema.TypeList,
-				Computed: true,
+				Type:      schema.TypeList,
+				Computed:  true,
 				Sensitive: true,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"api_version": {
-							Type: schema.TypeString,
-							Computed: true,
+							Type:      schema.TypeString,
+							Computed:  true,
 							Sensitive: true,
-
 						},
 						"current_context": {
-							Type: schema.TypeString,
-							Computed: true,
+							Type:      schema.TypeString,
+							Computed:  true,
 							Sensitive: true,
 						},
 						"kind": {
-							Type: schema.TypeString,
+							Type:     schema.TypeString,
 							Computed: true,
 						},
 						"users": {
-							Type: schema.TypeList,
+							Type:     schema.TypeList,
 							Computed: true,
 							Elem: &schema.Resource{
 								Schema: map[string]*schema.Schema{
 									"name": {
-										Type: schema.TypeString,
-										Computed: true,
+										Type:      schema.TypeString,
+										Computed:  true,
 										Sensitive: true,
-
 									},
 									"user": {
-										Type: schema.TypeMap,
-										Computed: true,
+										Type:      schema.TypeMap,
+										Computed:  true,
 										Sensitive: true,
 										Elem: &schema.Schema{
 											Type: schema.TypeString,
@@ -147,18 +145,18 @@ func dataSourceK8sCluster() *schema.Resource {
 							},
 						},
 						"clusters": {
-							Type: schema.TypeList,
+							Type:     schema.TypeList,
 							Computed: true,
 							Elem: &schema.Resource{
 								Schema: map[string]*schema.Schema{
 									"name": {
-										Type: schema.TypeString,
-										Computed: true,
+										Type:      schema.TypeString,
+										Computed:  true,
 										Sensitive: true,
 									},
 									"cluster": {
-										Type: schema.TypeMap,
-										Computed: true,
+										Type:      schema.TypeMap,
+										Computed:  true,
 										Sensitive: true,
 										Elem: &schema.Schema{
 											Type: schema.TypeString,
@@ -168,18 +166,18 @@ func dataSourceK8sCluster() *schema.Resource {
 							},
 						},
 						"contexts": {
-							Type: schema.TypeList,
+							Type:     schema.TypeList,
 							Computed: true,
 							Elem: &schema.Resource{
 								Schema: map[string]*schema.Schema{
 									"name": {
-										Type: schema.TypeString,
-										Computed: true,
+										Type:      schema.TypeString,
+										Computed:  true,
 										Sensitive: true,
 									},
 									"context": {
-										Type: schema.TypeMap,
-										Computed: true,
+										Type:      schema.TypeMap,
+										Computed:  true,
 										Sensitive: true,
 										Elem: &schema.Schema{
 											Type: schema.TypeString,
@@ -192,23 +190,23 @@ func dataSourceK8sCluster() *schema.Resource {
 				},
 			},
 			"user_tokens": {
-				Type: schema.TypeMap,
+				Type:      schema.TypeMap,
 				Sensitive: true,
-				Computed: true,
+				Computed:  true,
 				Elem: &schema.Schema{
-					Type: schema.TypeString,
+					Type:      schema.TypeString,
 					Sensitive: true,
 				},
 			},
 			"ca_crt": {
-				Type: schema.TypeString,
+				Type:      schema.TypeString,
 				Sensitive: true,
-				Computed: true,
+				Computed:  true,
 			},
 			"server": {
-				Type: schema.TypeString,
+				Type:      schema.TypeString,
 				Sensitive: true,
-				Computed: true,
+				Computed:  true,
 			},
 			"public": {
 				Type: schema.TypeBool,
@@ -356,7 +354,7 @@ func setK8sConfigData(d *schema.ResourceData, configStr string) error {
 		clustersList[i] = map[string]interface{}{
 			"name": cluster.Name,
 			"cluster": map[string]string{
-				"server": cluster.Cluster.Server,
+				"server":                     cluster.Cluster.Server,
 				"certificate_authority_data": string(decodedCrt),
 			},
 		}
@@ -370,7 +368,7 @@ func setK8sConfigData(d *schema.ResourceData, configStr string) error {
 			"name": contextVal.Name,
 			"context": map[string]string{
 				"cluster": contextVal.Context.Cluster,
-				"user": contextVal.Context.User,
+				"user":    contextVal.Context.User,
 			},
 		}
 	}

--- a/ionoscloud/resource_datacenter.go
+++ b/ionoscloud/resource_datacenter.go
@@ -250,17 +250,10 @@ func resourceDatacenterDelete(ctx context.Context, d *schema.ResourceData, meta 
 	return nil
 }
 
-func getImage(ctx context.Context, client *ionoscloud.APIClient, dcId string, imageName string, imageType string) (*ionoscloud.Image, error) {
+func resolveVolumeImageName(ctx context.Context, client *ionoscloud.APIClient, imageName string, location string) (*ionoscloud.Image, error) {
 
 	if imageName == "" {
 		return nil, fmt.Errorf("imageName not suplied")
-	}
-
-	dc, _, err := client.DataCenterApi.DatacentersFindById(ctx, dcId).Execute()
-
-	if err != nil {
-		log.Print(fmt.Errorf("error while fetching a data center ID %s %s", dcId, err))
-		return nil, err
 	}
 
 	images, _, err := client.ImageApi.ImagesGet(ctx).Execute()
@@ -277,15 +270,11 @@ func getImage(ctx context.Context, client *ionoscloud.APIClient, dcId string, im
 				imgName = *i.Properties.Name
 			}
 
-			if imageType == "SSD" {
-				imageType = "HDD"
-			}
-
-			if imgName != "" && strings.Contains(strings.ToLower(imgName), strings.ToLower(imageName)) && *i.Properties.ImageType == imageType && *i.Properties.Location == *dc.Properties.Location {
+			if imgName != "" && strings.Contains(strings.ToLower(imgName), strings.ToLower(imageName)) && *i.Properties.ImageType == "HDD" && *i.Properties.Location == location {
 				return &i, err
 			}
 
-			if imgName != "" && strings.ToLower(imageName) == strings.ToLower(*i.Id) && *i.Properties.ImageType == imageType && *i.Properties.Location == *dc.Properties.Location {
+			if imgName != "" && strings.ToLower(imageName) == strings.ToLower(*i.Id) && *i.Properties.ImageType == "HDD" && *i.Properties.Location == location {
 				return &i, err
 			}
 

--- a/ionoscloud/resource_datacenter.go
+++ b/ionoscloud/resource_datacenter.go
@@ -4,13 +4,11 @@ import (
 	"context"
 	"fmt"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
 	ionoscloud "github.com/ionos-cloud/sdk-go/v5"
 	"log"
 	"regexp"
-	"strings"
-
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
 func resourceDatacenter() *schema.Resource {
@@ -248,97 +246,6 @@ func resourceDatacenterDelete(ctx context.Context, d *schema.ResourceData, meta 
 
 	d.SetId("")
 	return nil
-}
-
-func resolveVolumeImageName(ctx context.Context, client *ionoscloud.APIClient, imageName string, location string) (*ionoscloud.Image, error) {
-
-	if imageName == "" {
-		return nil, fmt.Errorf("imageName not suplied")
-	}
-
-	images, _, err := client.ImageApi.ImagesGet(ctx).Execute()
-
-	if err != nil {
-		log.Print(fmt.Errorf("error while fetching the list of images %s", err))
-		return nil, err
-	}
-
-	if len(*images.Items) > 0 {
-		for _, i := range *images.Items {
-			imgName := ""
-			if i.Properties.Name != nil && *i.Properties.Name != "" {
-				imgName = *i.Properties.Name
-			}
-
-			if imgName != "" && strings.Contains(strings.ToLower(imgName), strings.ToLower(imageName)) && *i.Properties.ImageType == "HDD" && *i.Properties.Location == location {
-				return &i, err
-			}
-
-			if imgName != "" && strings.ToLower(imageName) == strings.ToLower(*i.Id) && *i.Properties.ImageType == "HDD" && *i.Properties.Location == location {
-				return &i, err
-			}
-
-		}
-	}
-	return nil, err
-}
-
-func getSnapshotId(ctx context.Context, client *ionoscloud.APIClient, snapshotName string) string {
-
-	if snapshotName == "" {
-		return ""
-	}
-
-	snapshots, _, err := client.SnapshotApi.SnapshotsGet(ctx).Execute()
-
-	if err != nil {
-		log.Print(fmt.Errorf("error while fetching the list of snapshots %s", err))
-	}
-
-	if len(*snapshots.Items) > 0 {
-		for _, i := range *snapshots.Items {
-			imgName := ""
-			if *i.Properties.Name != "" {
-				imgName = *i.Properties.Name
-			}
-
-			if imgName != "" && strings.Contains(strings.ToLower(imgName), strings.ToLower(snapshotName)) {
-				return *i.Id
-			}
-		}
-	}
-	return ""
-}
-
-func getImageAlias(ctx context.Context, client *ionoscloud.APIClient, imageAlias string, location string) string {
-
-	if imageAlias == "" {
-		return ""
-	}
-	parts := strings.SplitN(location, "/", 2)
-	if len(parts) != 2 {
-		log.Print(fmt.Errorf("invalid location id %s", location))
-	}
-
-	locations, _, err := client.LocationApi.LocationsFindByRegionIdAndId(ctx, parts[0], parts[1]).Execute()
-
-	if err != nil {
-		log.Print(fmt.Errorf("error while fetching the list of snapshots %s", err))
-	}
-
-	if len(*locations.Properties.ImageAliases) > 0 {
-		for _, i := range *locations.Properties.ImageAliases {
-			alias := ""
-			if i != "" {
-				alias = i
-			}
-
-			if alias != "" && strings.ToLower(alias) == strings.ToLower(imageAlias) {
-				return i
-			}
-		}
-	}
-	return ""
 }
 
 func IsValidUUID(uuid string) bool {

--- a/ionoscloud/resource_k8s_node_pool.go
+++ b/ionoscloud/resource_k8s_node_pool.go
@@ -468,6 +468,41 @@ func resourcek8sNodePoolUpdate(ctx context.Context, d *schema.ResourceData, meta
 		NodeCount: &nodeCount,
 	}
 
+	if d.HasChange("name") {
+		diags := diag.FromErr(fmt.Errorf("name attribute is immutable, therefore not allowed in update requests"))
+		return diags
+	}
+
+	if d.HasChange("cpu_family") {
+		diags := diag.FromErr(fmt.Errorf("cpu_family attribute is immutable, therefore not allowed in update requests"))
+		return diags
+	}
+
+	if d.HasChange("availability_zone") {
+		diags := diag.FromErr(fmt.Errorf("availability_zone attribute is immutable, therefore not allowed in update requests"))
+		return diags
+	}
+
+	if d.HasChange("cores_count") {
+		diags := diag.FromErr(fmt.Errorf("cores_count attribute is immutable, therefore not allowed in update requests"))
+		return diags
+	}
+
+	if d.HasChange("ram_size") {
+		diags := diag.FromErr(fmt.Errorf("ram_size attribute is immutable, therefore not allowed in update requests"))
+		return diags
+	}
+
+	if d.HasChange("storage_size") {
+		diags := diag.FromErr(fmt.Errorf("storage_size attribute is immutable, therefore not allowed in update requests"))
+		return diags
+	}
+
+	if d.HasChange("storage_type") {
+		diags := diag.FromErr(fmt.Errorf("storage_size attribute is immutable, therefore not allowed in update requests"))
+		return diags
+	}
+
 	if d.HasChange("k8s_version") {
 		oldk8sVersion, newk8sVersion := d.GetChange("k8s_version")
 		log.Printf("[INFO] k8s pool k8s version changed from %+v to %+v", oldk8sVersion, newk8sVersion)

--- a/ionoscloud/resource_server.go
+++ b/ionoscloud/resource_server.go
@@ -460,7 +460,7 @@ func resourceServerCreate(ctx context.Context, d *schema.ResourceData, meta inte
 				return diags
 			}
 
-			if *img.Properties.Public == true && isSnapshot == false {
+			if isSnapshot == false && img.Properties.Public != nil && *img.Properties.Public == true {
 
 				if volume.ImagePassword == nil && len(sshkeyPath) == 0 {
 					diags := diag.FromErr(fmt.Errorf("either 'image_password' or 'ssh_key_path' must be provided"))
@@ -494,7 +494,7 @@ func resourceServerCreate(ctx context.Context, d *schema.ResourceData, meta inte
 					isSnapshot = true
 				}
 
-				if img.Properties.Public != nil && *img.Properties.Public == true && isSnapshot == false {
+				if isSnapshot == false && img.Properties.Public != nil && *img.Properties.Public == true {
 					if volume.ImagePassword == nil && len(sshkeyPath) == 0 {
 						diags := diag.FromErr(fmt.Errorf("either 'image_password' or 'ssh_key_path' must be provided"))
 						return diags

--- a/ionoscloud/resource_server.go
+++ b/ionoscloud/resource_server.go
@@ -406,7 +406,13 @@ func resourceServerCreate(ctx context.Context, d *schema.ResourceData, meta inte
 
 	if imageName != "" {
 		if !IsValidUUID(imageName) {
-			img, err := getImage(ctx, client, dcId, imageName, *volume.Type)
+			dc, _, err := client.DataCenterApi.DatacentersFindById(ctx, dcId).Execute()
+			if err != nil {
+				diags := diag.FromErr(fmt.Errorf("error fetching datacenter %s: (%s)", dcId, err))
+				return diags
+			}
+
+			img, err := resolveVolumeImageName(ctx, client, imageName, *dc.Properties.Location)
 			if err != nil {
 				diags := diag.FromErr(err)
 				return diags
@@ -422,11 +428,6 @@ func resourceServerCreate(ctx context.Context, d *schema.ResourceData, meta inte
 					isSnapshot = true
 				} else {
 					log.Printf("[*****] lookig for an image alias for %s\n", imageName)
-					dc, _, err := client.DataCenterApi.DatacentersFindById(ctx, dcId).Execute()
-					if err != nil {
-						diags := diag.FromErr(fmt.Errorf("error fetching datacenter %s: (%s)", dcId, err))
-						return diags
-					}
 
 					imageAlias = getImageAlias(ctx, client, imageName, *dc.Properties.Location)
 					if imageAlias == "" {
@@ -466,7 +467,13 @@ func resourceServerCreate(ctx context.Context, d *schema.ResourceData, meta inte
 					return diags
 				}
 
-				img, err := getImage(ctx, client, d.Get("datacenter_id").(string), imageName, *volume.Type)
+				dc, _, err := client.DataCenterApi.DatacentersFindById(ctx, dcId).Execute()
+				if err != nil {
+					diags := diag.FromErr(fmt.Errorf("error fetching datacenter %s: (%s)", dcId, err))
+					return diags
+				}
+
+				img, err := resolveVolumeImageName(ctx, client, imageName, *dc.Properties.Location)
 
 				if err != nil {
 					diags := diag.FromErr(err)

--- a/ionoscloud/resource_server_test.go
+++ b/ionoscloud/resource_server_test.go
@@ -146,6 +146,26 @@ func TestAccServer_ResolveImageName(t *testing.T) {
 	})
 }
 
+func TestAccServer_WithSnapshot(t *testing.T) {
+	var server ionoscloud.Server
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+		ProviderFactories: testAccProviderFactories,
+		CheckDestroy:      testAccCheckServerDestroyCheck,
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(testAccCheckServerWithSnapshot),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckServerExists("ionoscloud_server.webserver", &server),
+				),
+			},
+		},
+	})
+}
+
 func testAccCheckServerDestroyCheck(s *terraform.State) error {
 	client := testAccProvider.Meta().(*ionoscloud.APIClient)
 
@@ -510,6 +530,69 @@ resource "ionoscloud_server" "webserver" {
     }
   }
 }`
+
+const testAccCheckServerWithSnapshot = `
+
+resource "ionoscloud_datacenter" "foobar" {
+	name       = "volume-test"
+	location   = "de/fra"
+}
+
+resource "ionoscloud_lan" "webserver_lan" {
+  datacenter_id = "${ionoscloud_datacenter.foobar.id}"
+  public = true
+  name = "public"
+}
+
+
+resource "ionoscloud_server" "webserver" {
+  name = "webserver"
+  datacenter_id = "${ionoscloud_datacenter.foobar.id}"
+  cores = 1
+  ram = 1024
+  availability_zone = "ZONE_1"
+  cpu_family = "INTEL_SKYLAKE"
+	image_name = "ubuntu:latest"
+	image_password = "K3tTj8G14a3EgKyNeeiY"
+  volume {
+    name = "system"
+    size = 5
+    disk_type = "SSD Standard"
+  }
+  nic {
+    lan = "${ionoscloud_lan.webserver_lan.id}"
+    dhcp = true
+    firewall_active = true
+  }
+}
+
+resource "ionoscloud_snapshot" "test_snapshot" {
+  datacenter_id = "${ionoscloud_datacenter.foobar.id}"
+  volume_id = "${ionoscloud_server.webserver.boot_volume}"
+  name = "terraform_snapshot"
+}
+
+resource "ionoscloud_server" "webserver2" {
+  depends_on = [ionoscloud_snapshot.test_snapshot]
+  name = "webserver2"
+  datacenter_id = "${ionoscloud_datacenter.foobar.id}"
+  cores = 1
+  ram = 1024
+  availability_zone = "ZONE_1"
+  cpu_family = "INTEL_SKYLAKE"
+  image_name = ionoscloud_snapshot.test_snapshot.id
+  volume {
+    name = "system"
+    size = 5
+    disk_type = "SSD Standard"
+  }
+  nic {
+    lan = "${ionoscloud_lan.webserver_lan.id}"
+    dhcp = true
+    firewall_active = true
+  }
+}
+`
 
 func Test_Update(t *testing.T) {
 

--- a/ionoscloud/resource_server_test.go
+++ b/ionoscloud/resource_server_test.go
@@ -122,6 +122,30 @@ func TestAccServer_NicIps(t *testing.T) {
 		},
 	})
 }
+
+func TestAccServer_ResolveImageName(t *testing.T) {
+	var server ionoscloud.Server
+	serverName := "webserver"
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+		ProviderFactories: testAccProviderFactories,
+		CheckDestroy:      testAccCheckServerDestroyCheck,
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(testAccCheckServerResolveImageName, serverName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckServerExists("ionoscloud_server.webserver", &server),
+					testAccCheckServerAttributes("ionoscloud_server.webserver", serverName),
+					resource.TestCheckResourceAttr("ionoscloud_server.webserver", "name", serverName),
+				),
+			},
+		},
+	})
+}
+
 func testAccCheckServerDestroyCheck(s *terraform.State) error {
 	client := testAccProvider.Meta().(*ionoscloud.APIClient)
 
@@ -434,6 +458,56 @@ resource "ionoscloud_server" "webserver" {
     dhcp            = true
     ips            = [ ionoscloud_ipblock.webserver_ipblock.ips[0], ionoscloud_ipblock.webserver_ipblock.ips[1] ]
     firewall_active = false
+  }
+}`
+
+const testAccCheckServerResolveImageName = `
+resource "ionoscloud_datacenter" "datacenter1" {
+  name        = "TF Test datacenter"
+  location    = "de/fra"
+  description = "Test datacenter done by TF"
+}
+
+resource "ionoscloud_lan" "public_lan" {
+  datacenter_id = ionoscloud_datacenter.datacenter1.id
+  public        = true
+}
+
+
+resource "ionoscloud_ipblock" "public_ip1" {
+  name     = "test-ip"
+  location = ionoscloud_datacenter.datacenter1.location
+  size     = 1
+}
+
+
+resource "ionoscloud_server" "webserver" {
+  name              = "%s"
+  datacenter_id     = ionoscloud_datacenter.datacenter1.id
+  cores             = 1
+  ram               = 1024
+  availability_zone = "ZONE_1"
+  cpu_family        = "INTEL_SKYLAKE" 
+  image_name        = "Ubuntu-20.04-LTS"
+  image_password    = "pass123456"
+  volume {
+    name           = "test1-root"
+    size              = 5
+    disk_type      = "SSD Standard"
+  }
+
+  nic {
+    lan             = ionoscloud_lan.public_lan.id
+    dhcp            = true
+    ips             = ionoscloud_ipblock.public_ip1.ips
+    firewall_active = true
+  
+    firewall {
+      protocol         = "TCP"
+      name             = "SSH"
+      port_range_start = 22
+      port_range_end   = 22
+    }
   }
 }`
 

--- a/ionoscloud/resource_server_test.go
+++ b/ionoscloud/resource_server_test.go
@@ -462,28 +462,28 @@ resource "ionoscloud_server" "webserver" {
 }`
 
 const testAccCheckServerResolveImageName = `
-resource "ionoscloud_datacenter" "datacenter1" {
-  name        = "TF Test datacenter"
+resource "ionoscloud_datacenter" "datacenter" {
+  name        = "test_server"
   location    = "de/fra"
   description = "Test datacenter done by TF"
 }
 
 resource "ionoscloud_lan" "public_lan" {
-  datacenter_id = ionoscloud_datacenter.datacenter1.id
+  datacenter_id = ionoscloud_datacenter.datacenter.id
   public        = true
 }
 
 
 resource "ionoscloud_ipblock" "public_ip1" {
   name     = "test-ip"
-  location = ionoscloud_datacenter.datacenter1.location
+  location = ionoscloud_datacenter.datacenter.location
   size     = 1
 }
 
 
 resource "ionoscloud_server" "webserver" {
   name              = "%s"
-  datacenter_id     = ionoscloud_datacenter.datacenter1.id
+  datacenter_id     = ionoscloud_datacenter.datacenter.id
   cores             = 1
   ram               = 1024
   availability_zone = "ZONE_1"

--- a/ionoscloud/resource_volume.go
+++ b/ionoscloud/resource_volume.go
@@ -155,7 +155,14 @@ func resourceVolumeCreate(ctx context.Context, d *schema.ResourceData, meta inte
 	var image string
 	if imageAlias == "" && imageName != "" {
 		if !IsValidUUID(imageName) {
-			img, err := getImage(ctx, client, dcId, imageName, d.Get("disk_type").(string))
+
+			dc, _, err := client.DataCenterApi.DatacentersFindById(ctx, dcId).Execute()
+			if err != nil {
+				diags := diag.FromErr(fmt.Errorf("error fetching datacenter %s: (%s)", dcId, err))
+				return diags
+			}
+
+			img, err := resolveVolumeImageName(ctx, client, imageName, *dc.Properties.Location)
 			if err != nil {
 				diags := diag.FromErr(err)
 				return diags
@@ -169,12 +176,6 @@ func resourceVolumeCreate(ctx context.Context, d *schema.ResourceData, meta inte
 				if image != "" {
 					isSnapshot = true
 				} else {
-					dc, _, err := client.DataCenterApi.DatacentersFindById(ctx, dcId).Execute()
-
-					if err != nil {
-						diags := diag.FromErr(fmt.Errorf("an error occured while fetching a Datacenter ID %s %s", dcId, err))
-						return diags
-					}
 					imageAlias = getImageAlias(ctx, client, imageName, *dc.Properties.Location)
 				}
 			}

--- a/ionoscloud/resource_volume.go
+++ b/ionoscloud/resource_volume.go
@@ -7,6 +7,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
 	ionoscloud "github.com/ionos-cloud/sdk-go/v5"
 	"log"
+	"strings"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
@@ -660,4 +661,95 @@ func resourceVolumeDelete(ctx context.Context, d *schema.ResourceData, meta inte
 
 	d.SetId("")
 	return nil
+}
+
+func resolveVolumeImageName(ctx context.Context, client *ionoscloud.APIClient, imageName string, location string) (*ionoscloud.Image, error) {
+
+	if imageName == "" {
+		return nil, fmt.Errorf("imageName not suplied")
+	}
+
+	images, _, err := client.ImageApi.ImagesGet(ctx).Execute()
+
+	if err != nil {
+		log.Print(fmt.Errorf("error while fetching the list of images %s", err))
+		return nil, err
+	}
+
+	if len(*images.Items) > 0 {
+		for _, i := range *images.Items {
+			imgName := ""
+			if i.Properties.Name != nil && *i.Properties.Name != "" {
+				imgName = *i.Properties.Name
+			}
+
+			if imgName != "" && strings.Contains(strings.ToLower(imgName), strings.ToLower(imageName)) && *i.Properties.ImageType == "HDD" && *i.Properties.Location == location {
+				return &i, err
+			}
+
+			if imgName != "" && strings.ToLower(imageName) == strings.ToLower(*i.Id) && *i.Properties.ImageType == "HDD" && *i.Properties.Location == location {
+				return &i, err
+			}
+
+		}
+	}
+	return nil, err
+}
+
+func getSnapshotId(ctx context.Context, client *ionoscloud.APIClient, snapshotName string) string {
+
+	if snapshotName == "" {
+		return ""
+	}
+
+	snapshots, _, err := client.SnapshotApi.SnapshotsGet(ctx).Execute()
+
+	if err != nil {
+		log.Print(fmt.Errorf("error while fetching the list of snapshots %s", err))
+	}
+
+	if len(*snapshots.Items) > 0 {
+		for _, i := range *snapshots.Items {
+			imgName := ""
+			if *i.Properties.Name != "" {
+				imgName = *i.Properties.Name
+			}
+
+			if imgName != "" && strings.Contains(strings.ToLower(imgName), strings.ToLower(snapshotName)) {
+				return *i.Id
+			}
+		}
+	}
+	return ""
+}
+
+func getImageAlias(ctx context.Context, client *ionoscloud.APIClient, imageAlias string, location string) string {
+
+	if imageAlias == "" {
+		return ""
+	}
+	parts := strings.SplitN(location, "/", 2)
+	if len(parts) != 2 {
+		log.Print(fmt.Errorf("invalid location id %s", location))
+	}
+
+	locations, _, err := client.LocationApi.LocationsFindByRegionIdAndId(ctx, parts[0], parts[1]).Execute()
+
+	if err != nil {
+		log.Print(fmt.Errorf("error while fetching the list of snapshots %s", err))
+	}
+
+	if len(*locations.Properties.ImageAliases) > 0 {
+		for _, i := range *locations.Properties.ImageAliases {
+			alias := ""
+			if i != "" {
+				alias = i
+			}
+
+			if alias != "" && strings.ToLower(alias) == strings.ToLower(imageAlias) {
+				return i
+			}
+		}
+	}
+	return ""
 }

--- a/ionoscloud/resource_volume.go
+++ b/ionoscloud/resource_volume.go
@@ -713,6 +713,7 @@ func getSnapshotId(ctx context.Context, client *ionoscloud.APIClient, snapshotNa
 			imgName := ""
 			if *i.Properties.Name != "" {
 				imgName = *i.Properties.Name
+				fmt.Printf("snapshot name: %s \n\n ", *i.Properties.Name)
 			}
 
 			if imgName != "" && strings.Contains(strings.ToLower(imgName), strings.ToLower(snapshotName)) {

--- a/ionoscloud/resource_volume.go
+++ b/ionoscloud/resource_volume.go
@@ -185,7 +185,7 @@ func resourceVolumeCreate(ctx context.Context, d *schema.ResourceData, meta inte
 				diags := diag.FromErr(fmt.Errorf("could not find an image/imagealias/snapshot that matches %s ", imageName))
 				return diags
 			}
-			if imagePassword == "" && len(sshKeypath) == 0 && isSnapshot == false && img != nil && *img.Properties.Public {
+			if imagePassword == "" && len(sshKeypath) == 0 && isSnapshot == false && img != nil && img.Properties.Public != nil && *img.Properties.Public {
 				diags := diag.FromErr(fmt.Errorf("either 'image_password' or 'sshkey' must be provided"))
 				return diags
 			}
@@ -198,7 +198,7 @@ func resourceVolumeCreate(ctx context.Context, d *schema.ResourceData, meta inte
 					return diags
 				}
 				isSnapshot = true
-			} else if *img.Properties.Public == true && isSnapshot == false {
+			} else if isSnapshot == false && img.Properties.Public != nil && *img.Properties.Public == true {
 				if imagePassword == "" && len(sshKeypath) == 0 {
 					diags := diag.FromErr(fmt.Errorf("either 'image_password' or 'sshkey' must be provided"))
 					return diags

--- a/ionoscloud/resource_volume_test.go
+++ b/ionoscloud/resource_volume_test.go
@@ -59,6 +59,27 @@ func TestAccVolume_NoPassword(t *testing.T) {
 	})
 }
 
+func TestAccVolume_ResolveImageName(t *testing.T) {
+	var volume ionoscloud.Volume
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+		ProviderFactories: testAccProviderFactories,
+		CheckDestroy:      testAccCheckVolumeDestroyCheck,
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(testAccCheckVolumeResolveImageName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckVolumeExists("ionoscloud_volume.image_name_volume", &volume),
+					resource.TestCheckResourceAttr("ionoscloud_volume.image_name_volume", "name", "image_name_volume"),
+				),
+			},
+		},
+	})
+}
+
 func testAccCheckVolumeDestroyCheck(s *terraform.State) error {
 	client := testAccProvider.Meta().(*ionoscloud.APIClient)
 
@@ -253,3 +274,49 @@ resource "ionoscloud_volume" "no_password_volume" {
   disk_type      = "HDD"
   licence_type   =  "other"
 }`
+
+const testAccCheckVolumeResolveImageName = `
+resource "ionoscloud_datacenter" "foobar" {
+	name       = "volume-test"
+	location   = "de/fra"
+}
+
+resource "ionoscloud_lan" "webserver_lan" {
+  datacenter_id = "${ionoscloud_datacenter.foobar.id}"
+  public = true
+  name = "public"
+}
+
+resource "ionoscloud_server" "webserver" {
+  name = "webserver"
+  datacenter_id = "${ionoscloud_datacenter.foobar.id}"
+  cores = 1
+  ram = 1024
+  availability_zone = "ZONE_1"
+  cpu_family = "INTEL_SKYLAKE"
+	image_name = "ubuntu:latest"
+	image_password = "K3tTj8G14a3EgKyNeeiY"
+  volume {
+    name = "system"
+    size = 5
+    disk_type = "SSD Standard"
+  }
+  nic {
+    lan = "${ionoscloud_lan.webserver_lan.id}"
+    dhcp = true
+    firewall_active = true
+  }
+}
+
+resource "ionoscloud_volume" "image_name_volume" {
+  datacenter_id = "${ionoscloud_datacenter.foobar.id}"
+  server_id = "${ionoscloud_server.webserver.id}"
+  availability_zone = "ZONE_1"
+  name = "image_name_volume"
+  size = 5
+  disk_type = "SSD Standard"
+  bus = "VIRTIO"
+  image_name = "Ubuntu-20.04-LTS"
+  image_password = "K3tTj8G14a3EgKyNeeiY"
+}
+`

